jameel-institute/daedalus.api
# jidea-74 Get model result via rrq

Adds [rrq](https://github.com/mrc-ide/rrq) to the api so that all model run requests are run in queued tasks. The `scenario/run`,  `scenario/status` and `scenario/results` endpoints are now fully implemented -  they spawn the task, report on its status and fetch its results respectively. A separate `model_run` method is the implementation of the task, and for now this is still just reading and returning the dummy data .

This implementation is largely based on that in orderly.runner. There is a `Queue` class which manages the rrq interactions, configuring the queue, and dealing with tasks. 

We need (at least) two additional docker containers to support this pattern - one for redis (which is used by rrq to manage the queue) and at least one more for the worker(s) which execute the tasks on the queue. 

Both the main api container and the worker(s) will run the same daedalus.api image, but should be spun up with different entry points: `main` for the api, `main_worker` for the workers. 

I've modified the test connection script a bit to pull out the running of the containers so that you can do that separately for manual testing.

### To test "running" a model:
1. Run the containers with `./docker/run_containers`
2. Post a request to run a model: 
``` 
curl -H "Content-Type: application/json" -d '{"modelVersion": "0.0.1", "parameters": {}}' http://localhost:8001/scenario/run
```
You should get a successful response which includes a runId.

3. Check the status of the job, replacing `[RUN_ID]` with the runId received in 2.
```
curl http://localhost:8001/scenario/status/[RUN_ID]
```
You should see a status result indicating that the job completed successfully. 

4. Get the results of the job:
```
curl http://localhost:8001/scenario/results/[RUN_ID]
```
You should see the hardcoded dummy model results returned. 

5. You can tear down the containers with `./docker/clear_containers`


### Notes
I've modified the status and results routes to take the run id as the final part of the route rather than as a query string param, as that seems a little nicer/more RESTful. 

I found that I had to downgrade the base docker image (to the version known to work for orderly.runner), as I was getting the following error: `unable to load shared object '/usr/local/lib/R/site-library/openssl/libs/openssl.so':\n  libssl.so.1.1: cannot open shared object file: No such file or directory.` - even though the shared object file did exist at the path shown. 

I couldn't crack getting mocking to work inside the Queue constructor, so the queue unit tests only mock  within the other methods.
